### PR TITLE
Using mermaid for class diagrams

### PR DESCRIPTION
**Pull request by TotoGaz**
Using mermaid for some UML diagrams (see e.g. https://geosx-geosx--1897.com.readthedocs.build/en/1897/coreComponents/mesh/docs/meshDeveloperGuide.html)

Given some critical lacks in `mermaid` (no `package/namespace`, no easy coloring in the mardown/sphinx. no generic functions...), I think that we should use both `plantuml` and `mermaid`.
Let's favor `mermaid` for simple cases, (and because hopefully this tool will improve given the visibility provided by github). But when it's getting more complicated, let's use `plantuml`.
Converting from one to the other is not a big deal.

### CODE REVIEW

**Comment by rrsettgast** [comment]
@TotoGaz Can you explain what you would like to use the diagrams for?

**Comment by TotoGaz** [comment]
> @TotoGaz Can you explain what you would like to use the diagrams for?

Sometimes it's convenient to explain some architectural concepts with diagram and the selected level of information.
E.g. in https://geosx-geosx--1892.com.readthedocs.build/en/1892/coreComponents/constitutive/PVTPackage/docs/main.html it's easier to have a global understanding of the PVT package.

Selecting what to show for which purpose is not something one can easily delegate to `doxygen`.
E.g. we can better explain what's around `CellBlockManager` with this kind of diagrams; new developers could appreciate this kind of tools.

File: src/conf.py [hunk 1]
```diff
@@ -99,9 +99,15 @@
     'sphinx.ext.doctest',
     'sphinx.ext.inheritance_diagram',
     'matplotlib.sphinxext.plot_directive',
-    'sphinx.ext.napoleon'
+    'sphinx.ext.napoleon',
+    'sphinxcontrib.mermaid',
 ]
-plot_html_show_source_link = True
```
Comment (herve-gross) [inline src/conf.py:104]: Did you intend to remove this?

File: src/coreComponents/mesh/docs/meshDeveloperGuide.rst [hunk 2]
```diff
@@ -12,12 +12,9 @@ how instantiations of each class relate to each other in the data hierarchy rath
 type relates to each other in an inheritance diagram.
 
 .. _diagMeshDevFig:
-.. figure:: /coreComponents/mesh/docs/MeshObjectInstantiationHierarchy.png
-    :align: center
-    :figclass: align-center
-    :name: MeshObjectInstantiationHierarchy
+.. mermaid:: /coreComponents/mesh/docs/MeshObjectInstantiationHierarchy.mmd
+   :caption: Instances cardinalities and multiplicities mesh domain.
```
[the remark this answers is not in the document]
Comment (TotoGaz) [reply src/coreComponents/mesh/docs/meshDeveloperGuide.rst:16]: What about "_Object instances for the mesh domain. Cardinalities and relationships are indicated._"?

**Comment by TotoGaz** [comment]
> @TotoGaz I am not saying that this is better, but Doxygen creates uml diagrams for you if you would like. Of course, they can get a little messy, but they are automatic.

The limitation of `doxygen` in such a case is that you cannot really select what you think should be exposed in the diagram.
E.g. in https://geosx-geosx--1892.com.readthedocs.build/en/1892/_images/plantuml-60c4eacfb1755e523b8b57587fb0e6cab089a165.svg I've selected what I considered to be the proper level of collaboration, inheritance, coloring, members,... such that, combined with the appropriate level of documentation (still to come 🤣 ), a new dev gets the big picture of the package.

Another example in 
```mermaid
classDiagram

class DomainPartition {
    +getMeshBodies() List~MeshBody~
}

class MeshBody {
    +getMeshLevels() List~MeshLevels~
}

class MeshLevel {
    +getNodeManager() NodeManager
    +getEdgeManager() EdgeManager
    +getFaceManager() FaceManager
    +getElemManager() ElementRegionManager
}

class NodeManager {
    +edgeList()
    +faceList()
    +elementList()
    +elementSubRegionList()
    +elementRegionList()
}

class EdgeManager {
    +nodeList()
    +faceList()
}

class FaceManager {
    +nodeList()
    +edgeList()
    +elementList()
    +elementSubRegionList()
    +elementRegionList()
}

class ElementRegionManager {
    +forElementRegion(f)
    +forElementSubRegion(f)
}

DomainPartition "1" *-- "1..*" MeshBody
MeshBody "1" *-- "1..*" MeshLevel

MeshLevel "1" *-- "1" NodeManager
MeshLevel "1" *-- "1" EdgeManager
MeshLevel "1" *-- "1" FaceManager
MeshLevel "1" *-- "1" ElementRegionManager

ElementRegionManager "1" *-- "0..n" WellElementRegion
ElementRegionManager "1" *-- "0..n" FaceElementRegion
ElementRegionManager "1" *-- "0..n" CellElementRegion

WellElementRegion "1" *-- "1..n" WellElementSubRegion
FaceElementRegion "1" *-- "1..n" FaceElementSubRegion
CellElementRegion "1" *-- "1..n" CellElementSubRegion
```
I've added multiplicities which is not managed by `doxygen` (I think). I've decided not to expose the `pack` member functions because it's not the subject of the current doc... But I've considered the members that made sense in this context.
I also took freedom to indicate that `DomainPartition::DomainPartition()` returns a `List< MeshBody >`, which it does not exactly, because it's _practically_ equivalent and discussing the `Group` concept here could hinder understanding.

If I ever need to use `FaceManager` in another context, I'll have to build another `doxygen` configuration? That sounds tricky.
So yes there is a little consistency to be maintained by hand, but I think that's a fair price w.r.t. the benefits.

Hopefully soon we'll have the same about `CellBlockManager*` which will help understand this part of the code too.

**Review by herve-gross — APPROVED** [review]
Looks good to me, I hope we will use and maintain mermaid documentation for high-touch areas of the code.

**Pull request merged**